File: android2_lesson06/src/java/java/com/example/ama/android2_lesson06/ui/base/BaseSmsPresenter.java
package com.example.ama.android2_lesson06.ui.base;

import com.example.ama.android2_lesson06.repo.SmsStorageManager;

import io.reactivex.disposables.CompositeDisposable;

public abstract class BaseSmsPresenter<T extends SmsExampleView> implements SmsPresenter<T> {

    protected T view;
    protected SmsStorageManager manager;
    protected CompositeDisposable disposable;

    @Override
    public void attachView(T view) {
        this.view = view;
        if (disposable == null) {
            disposable = new CompositeDisposable();
        }
        if (manager == null) {
            manager = new SmsStorageManager();
        }
    }

    @Override
    public void detachView() {
        view = null;
        if (disposable != null) {
            disposable.clear();
        }
    }
}

File: android2_lesson06/src/kotlin/java/com/example/ama/android2_lesson06/ui/base/BaseSmsPresenter.kt
package com.example.ama.android2_lesson06.ui.base

import com.example.ama.android2_lesson06.repo.SmsStorageManager
import io.reactivex.disposables.CompositeDisposable

abstract class BaseSmsPresenter<T : SmsExpampleView> : SmsPresenter<T> {

    protected var view: T? = null
    protected var manager: SmsStorageManager = SmsStorageManager()
    protected var disposable: CompositeDisposable? = null

    override fun attachView(view: T) {
        this.view = view
        if (disposable == null) {
            disposable = CompositeDisposable()
        }
    }

    override fun detachView() {
        view = null
        disposable?.clear()
    }
}